### PR TITLE
QKeras support for RNN layers

### PR DESCRIPTION
**Pull request by laurilaatu**
# Description

> :memo: This pull request adds QKeras support for RNN layers

The quantizers for kernel, recurrent kernel and bias are retrieved from the QKeras model. State quantizer corresponds to the output precision of the hidden state which is set by altering the default precision (discussed in issue #825). Currently the quantized activations for the RNN layers are not supported.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>
> * Tested with several different network configurations
> * Pytests implemented
> * All layers produce expected output except for the Vivado implementation of the SimpleRNN (excluded from pytests)

**Test Configuration**: Quartus 21.1.0.169.pro

Example usage:
```
qmodel = Sequential()
qmodel.add(QSimpleRNN(4, input_shape=(5,1), name='rnn',
                      kernel_quantizer=quantized_bits(bits=9, integer=0, symmetric=False, alpha=1.0),
                      recurrent_quantizer=quantized_bits(bits=9, integer=0, symmetric=False, alpha=1.0),
                      bias_quantizer=quantized_bits(bits=9, integer=0, symmetric=False, alpha=1.0),
                      state_quantizer=quantized_bits(bits=9, integer=0, symmetric=False, alpha=1.0),
                      activation='relu'))
qmodel.summary()
qmodel.compile(loss='mse', optimizer='adam')


config = hls4ml.utils.config_from_keras_model(qmodel,
                                           default_precision="ap_fixed<9,1>",
                                           granularity='name')

hls_model = hls4ml.converters.convert_from_keras_model(qmodel,
                                           hls_config=config,
                                           output_dir='model_qrnn/hls4ml_prj',
                                           part='AGFB014R24A2E2VR0',
                                           backend='Quartus')
hls_model.compile()
```

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/main/CONTRIBUTING.md).
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have installed and run `pre-commit` on the files I edited or added.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by jmitrevs** [comment]
@laurilaatu , is it possible for you to fix the conflicts so we can merge this PR?

**Comment by laurilaatu** [comment]
@jmitrevs , the conflict is now resolved.

**Comment by vloncar** [comment]
The newly added test [fails](https://gitlab.cern.ch/fastmachinelearning/hls4ml/-/jobs/33238566) with:
```
FAILED test_qkeras.py::test_qsimplernn[Quartus] - AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0.1
Mismatched elements: 1 / 4 (25%)
Max absolute difference: 0.953125
Max relative difference: 1.
 x: array([[0.      , 1.004456, 0.024292, 0.352661]], dtype=float32)
 y: array([[0.953125, 0.921875, 0.015625, 0.351562]])
```
@laurilaatu can you investigate?

**Comment by laurilaatu** [comment]
@vloncar it seems that the test fails occasionally due to the random initialization. The latest commit fixes the issue by setting the seed. The issue does not affect `QLSTM` or `QGRU`.

Please let me know if any other changes are required.

File: test/pytest/test_qkeras.py [hunk 1]
```diff
@@ -480,6 +481,109 @@ def test_quantised_po2_bit_width(backend, io_type, strategy):
     np.testing.assert_allclose(y_hls.flatten(), y_keras.flatten(), rtol=2e-2)
 
 
+@pytest.mark.parametrize('backend', ['Quartus'])
+def test_qsimplernn(backend):
+    '''
+    Test proper handling of QSimpleRNN.
+    '''
+    set_random_seed(42)
```
Comment (jmitrevs) [inline test/pytest/test_qkeras.py:489]: Generally you should not set the seed. We use pytest-randomly to set the seed. If there is sensitivity to random numbers, maybe the bitwidths or rounding/saturation need to be changed?

Comment (laurilaatu) [reply test/pytest/test_qkeras.py]: Fixed by increasing the bitwidth and reducing the `linspace` range .

**Review by jmitrevs — APPROVED** [review]

**Pull request merged**